src/store/image-stats.ts

```diff
@@ -0,0 +1,223 @@
+import { HistogramWorker } from '@/src/utils/histogram.worker';
+import { Maybe } from '@/src/types';
+import { useImage } from '@/src/composables/useCurrentImage';
+import { useImageCacheStore } from './image-cache';
+import { useMessageStore } from './messages';
+
+export type ImageStats = {
+  scalarMin: number;
+  scalarMax: number;
+  autoRangeValues?: Record<string, [number, number]>;
+};
+
+async function computeAutoRangeValues(
+  imageData: vtkImageData,
+  isImageLoading: boolean
+): Promise<Record<string, [number, number]>> {
+  if (isImageLoading || !imageData) {
+    return {};
+  }
+
+  const scalars = imageData.getPointData()?.getScalars();
+  if (!scalars) {
+    return {};
+  }
+
+  const worker = Comlink.wrap<HistogramWorker>(
+    new Worker(new URL('@/src/utils/histogram.worker.ts', import.meta.url), {
+      type: 'module',
+    })
+  );
+
+  const scalarData = scalars.getData() as number[];
+  const { min, max } = vtkDataArray.fastComputeRange(scalarData, 0, 1);
+  const hist = await worker.histogram(scalarData, [min, max], WL_HIST_BINS);
+  worker[Comlink.releaseProxy]();
+
+  const cumulativeHist: number[] = [];
+  hist.reduce((acc, val) => {
+    const currentSum = acc + val;
+    cumulativeHist.push(currentSum);
+    return currentSum;
+  }, 0);
+
+  const width = (max - min + 1) / WL_HIST_BINS;
+  const totalCount = scalarData.length;
+
+  return Object.fromEntries(
+    Object.entries(WLAutoRanges).map(([key, percentage]) => {
+      const lowerBound = percentage * 0.01 * totalCount;
+      const upperBound = (1 - percentage * 0.01) * totalCount;
+
+      const startIdx = cumulativeHist.findIndex((v) => v >= lowerBound);
+      const endIdx = cumulativeHist.findIndex((v) => v >= upperBound);
+
+      const start = Math.max(min, min + width * startIdx);
+      const end = Math.min(max, min + width * (endIdx + 1)); // Adjusted end calculation
+      return [key, [start, end] as [number, number]];
+    })
+  );
+}
+
+export const useImageStatsStore = defineStore('image-stats', () => {
+  const stats = reactive<Record<string, ImageStats>>({});
+  const imageCacheStore = useImageCacheStore();
+  const messageStore = useMessageStore();
+
+  const scalarRangeWatchers: Record<string, () => void> = {};
+  const autoRangeComputations: Record<
+    string,
+    Promise<Record<string, [number, number]>>
+  > = {};
+
+  const internalSetScalarRange = (
+    imageID: string,
+    min: number,
+    max: number
+  ) => {
+    stats[imageID] = {
+      ...stats[imageID],
+      scalarMin: min,
+      scalarMax: max,
+    };
+  };
+
+  const internalSetAutoRangeValues = (
+    imageID: string,
+    autoValues: Record<string, [number, number]>
+  ) => {
+    stats[imageID] = {
+      ...(stats[imageID] ?? { scalarMin: 0, scalarMax: 0 }),
+      autoRangeValues: autoValues,
+    };
+  };
+
+  const internalRemoveStats = (imageID: string) => {
+    delete stats[imageID];
+  };
+
+  const setupImageWatcher = (id: string) => {
```

Comment:
This function has multiple watchers. I would suggest using an effectScope so you can clean everything up at once.

Reply:
Added an EffectScope in the code that calls this function.